src/components/mask-input/mask-input.ts

```diff
@@ -0,0 +1,360 @@
+      end = this.inputSelection.end;
+    this.updateInput(data, start, end);
+  }
+
+  protected handleInput({ inputType, isComposing }: InputEvent) {
+    const value = this.input.value;
+    const start = this.selection.start;
+    let end = this.selection.end;
+
+    switch (inputType) {
+      case 'deleteContentForward':
+        this.updateInput('', start, (end = start === end ? ++end : end));
+        return this.updateComplete.then(() =>
+          this.input.setSelectionRange(end, end)
+        );
+
+      case 'deleteContentBackward':
+        if (isComposing) return;
+        return this.updateInput('', this.inputSelection.start, end);
+
+      case 'deleteByCut':
+        return this.updateInput('', start, end);
+
+      case 'insertText':
+        return this.updateInput(
+          value.substring(start, this.inputSelection.end),
+          start,
+          end
+        );
+
+      case 'insertFromPaste':
+        return this.updateInput(
+          value.substring(start, this.inputSelection.end),
+          start,
+          this.inputSelection.start
+        );
+
+      case 'insertFromDrop':
+        return this.insertFromDrop(this.input.value);
+    }
+  }
+
+  protected insertFromDrop(value: string) {
+    const { start, end } = this.inputSelection;
+    this.maskedValue = this.parser.apply(value);
+    this._value = this.parser.parse(this.maskedValue);
+    this.requestUpdate();
+    this.updateComplete.then(() => this.input.setSelectionRange(start, end));
+  }
+
+  protected updateInput(part: string, start: number, finish: number) {
+    const { value, end } = this.parser.replace(
+      this.maskedValue,
+      part,
+      start,
+      finish
+    );
+    this.maskedValue = value;
+    this._value = this.parser.parse(value);
+
+    this.requestUpdate();
+    if (start !== this.mask.length) {
+      this.emitEvent('igcInput', { detail: this.value });
+    }
+    this.updateComplete.then(() => this.input.setSelectionRange(end, end));
+  }
+
+  protected handleCut() {
+    this.selection = this.inputSelection;
+  }
+
+  protected handleDragStart() {
+    this.selection = this.inputSelection;
+  }
+
+  protected handleDragEnter() {
+    if (!this.hasFocus) {
+      this.maskedValue = this.parser.apply(this._value);
+    }
+  }
+
+  protected handleDragLeave() {
+    if (!this.hasFocus) {
+      this.updateMaskedValue();
+    }
+  }
+
+  protected override handleFocus() {
+    this.hasFocus = true;
+    this.maskedValue = this.parser.apply(this._value);
+    super.handleFocus();
+
+    // In case of empty value, select the whole mask
+    if (!this._value) {
+      this.updateComplete.then(() => this.select());
+    }
+  }
+
+  protected override handleBlur() {
+    this.hasFocus = false;
+    this.updateMaskedValue();
+    super.handleBlur();
+  }
+
+  protected override handleChange() {
+    this.emitEvent('igcChange', { detail: this.value });
+    this.invalid = !this.parser.isValidString(this._value);
```

Comment:
Since `parser.isValidString` will always return false for empty strings when there's a required position, it should only be checked if there is a value